must have full path to bash, otherwise fails (on Arch)

gdamjan opened the pull request.

gdamjan committed:
must have full path to bash, otherwise fails (on Arch)

Signed-off-by: Damjan Georgievski <gdamjan@gmail.com>

jfrazelle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

`git commit --amend -s --no-edit && git push -f`

gdamjan commented:
is it good now?

jfrazelle commented:
yes all good :D

jfrazelle commented:
@tianon is there anything that could go wrong with this?

tianon commented:
I use this script all the time and have never run into an issue with it.  Can you elaborate on your environment?

gdamjan commented:
I've digged a bit more into it.

Arch doesn't really have /bin or /sbin (they are symlinks now, per the /usr unification thing [1]) so it doesn't have them in the PATH either. the PATH variable is inherited from the host shell to the commands running in the chroot container. So it is `/usr/local/sbin:/usr/local/bin:/usr/bin` when run from default Arch system.

So one approach would be to specify full path to executables, and in the case of `bash` I would say it's better to use `/bin/bash`, but also that script should maybe normalize PATH to what Debian expects for normal operation (i.e. `export PATH=/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin`).

[1]
http://lwn.net/Articles/483921/

tianon commented:
Ah, thanks for digging; that makes more sense!  In that case, I think I'd rather see us actually `export PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin` somewhere in this script, especially to make sure that even though we've specified the full path to bash, we can't control things that `apt-get` might run, so I think we should be more defensive and provide the environment they're expecting.

gdamjan committed:
export a proper debian PATH for the chroot

when running in the chroot, it would be expected to have a Debian-ish environment,
most important the PATH variable. Otherwise the host PATH would be used which can wrong
for Debian, especially if the host is Arch (which doesn't have /bin, /sbin or /usr/sbin).

Signed-off-by: Damjan Georgievski <gdamjan@gmail.com>

crosbymichael commented:
LGTM

ping @tianon

tianon commented on contrib/mkimage/debootstrap:
```diff
@@ -189,9 +189,10 @@ fi
 	
 	# make sure we're fully up-to-date
-	chroot "$rootfsDir" bash -c 'apt-get update && apt-get dist-upgrade -y'
+	chroot "$rootfsDir" /bin/bash -c 'apt-get update && apt-get dist-upgrade -y'
```
Doesn't the full `PATH` mean we don't need this explicit `/bin/` anymore?

gdamjan replied:
true. 
I usually do prefer explicit paths for various reasons, but in this case it's the least of issues.

tianon commented:
FYI @gdamjan it doesn't notify us when you've updated the PR, so feel free to poke :+1:

jfrazelle commented:
@gdamjan are you able to make these changes otherwise one of us can carry the PR to make sure they get in :)

jfrazelle commented:
I believe this also fixes https://github.com/docker/docker/issues/8258 right?

tianon commented:
@jfrazelle indeed it does!

tianon commented:
I'm going to carry this one from here.
